File: app/src/main/java/agrawal/bhanu/jetpack/modules/AppModule.java
package agrawal.bhanu.jetpack.modules;

import android.app.Application;
import android.app.WallpaperManager;
import android.content.SharedPreferences;

import com.android.volley.RequestQueue;
import com.android.volley.toolbox.Volley;
import com.google.gson.Gson;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class AppModule {

    private Application mApplication;

    public AppModule(Application application){
        mApplication = application;
    }

    @Provides
    @Singleton
    public Application providesApplication(){
        return mApplication;
    }

    @Provides
    @Singleton
    public RequestQueue providesRequestQueue(Application application){
        return Volley.newRequestQueue(application.getApplicationContext());

    }


    @Provides
    @Singleton
    public WallpaperManager getWallpaperManager(Application application){
        return WallpaperManager.getInstance(application.getApplicationContext());
    }

    @Provides
    @Singleton
    public Gson providesGson(){
        return new Gson();
    }

    @Provides
    @Singleton
    public Executor providesExecuter(){
        return  Executors.newFixedThreadPool(5);
    }

    @Provides
    @Singleton
    public SharedPreferences provideSharedPreferences(Application application){
        return application.getSharedPreferences("launcher", application.MODE_PRIVATE);

    }
}

File: app/src/main/java/agrawal/bhanu/jetpack/modules/AppModule.kt
package agrawal.bhanu.jetpack.modules

import agrawal.bhanu.jetpack.SimpleIdlingResource
import android.app.Application
import android.app.WallpaperManager
import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import androidx.annotation.NonNull
import androidx.annotation.VisibleForTesting
import androidx.test.espresso.IdlingResource
import com.android.volley.RequestQueue
import com.android.volley.toolbox.Volley
import com.google.gson.Gson
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import javax.inject.Singleton

@InstallIn(SingletonComponent::class)
@Module
class AppModule {

    @Provides
    @Singleton
    fun provideUrlBuilder(): Uri.Builder {
        return Uri.Builder()
    }

    @Provides
    @Singleton
    fun getWallpaperManager(@ApplicationContext appContext: Context): WallpaperManager {
        return WallpaperManager.getInstance(appContext)
    }

    @Provides
    @Singleton
    fun providesGson() = Gson()

    @Provides
    @Singleton
    fun providesExecuter(): Executor {
        return Executors.newFixedThreadPool(5)
    }

    @Provides
    fun provideSharedPreferences(@ApplicationContext appContext: Context): SharedPreferences {
        return appContext.getSharedPreferences("launcher", Context.MODE_PRIVATE)
    }


    /**
     * Only called from test, creates and returns a new [SimpleIdlingResource].
     */
    @Provides
    @Singleton
    fun provideIdlingResource(): SimpleIdlingResource = SimpleIdlingResource()
}